### PR TITLE
Fixes determinant gradient for 1 x 1 matrices

### LINKED ISSUE #80761: `det` will return wrong gradient for `1x1` matrix with 0 value.
### 🐛 Describe the bug

`det` will return wrong gradient for `1x1` matrix with 0 value.

```python
import torch

input = torch.tensor([[0.]], dtype=torch.float64, requires_grad=True)
torch.det(input).backward()
print(input.grad)
# tensor([[0.]], dtype=torch.float64)
```

The correct gradient should be 1. Instead, when the value isn't zero, it will return the correct gradient.

```python
import torch

input = torch.tensor([[0.1]], dtype=torch.float64, requires_grad=True)
torch.det(input).backward()
print(input.grad)
# tensor([[1.]], dtype=torch.float64)
```

### Versions

pytorch: 1.11.0

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by redwrasse**
Fixes https://github.com/pytorch/pytorch/issues/80761, by adding special case handling for 1 x 1 matrices, to ensure gradient of the determinant is 1 for all input values. Also adds a regression test.

Note, as a separate issue, it might also be worth special casing 0 x 0 matrices as well to ensure gradient is 0 for the determinant function, since `d(det(A))/dA_ij = C_ij`, where `C_ij` is the cofactor matrix, and `C_ij` is implicitly assumed to be 1 for 0 x 0 matrices in the recursive determinant expression `det(A) = sum_k A_ik C_ik`- hence gradient of a constant is 0. This would also be consistent with other torch matrix operation gradients on 0 x 0 matrices:

```
# All other operations return tensor([], size=(0, 0))
torch.sum(empty_matrix).backward()    → grad: tensor([], size=(0, 0)) ✓
torch.trace(empty_matrix).backward()  → grad: tensor([], size=(0, 0)) ✓
torch.det(empty_matrix).backward()    → grad: None                     ✗
```

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -1038,6 +1038,14 @@ def test_det(self, device, dtype):
         with self.assertRaises(RuntimeError):
             op(t)
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.double, torch.cdouble)
+    def test_det_backward(self):
+        # Regression test for #80761.
+        input = torch.tensor([[0.]], dtype=torch.float64, requires_grad=True)
+        self.assertTrue(torch.autograd.gradcheck(torch.det, inputs=(input)))
```
Comment (zou3519) [inline test/test_linalg.py]: (input) isn't actually a tuple, what are you trying to do here

Comment (redwrasse) [reply test/test_linalg.py]: Good catch, removed the redundant parentheses.

This is just a regression test for the original bug (https://github.com/pytorch/pytorch/issues/80761#issue-1291018891): derivative at 0 of the determinant is 1.

Comment (redwrasse) [reply test/test_linalg.py]: @zou3519, fixed the syntax issue.^

**Review by Skylion007 — APPROVED** [review]
LGTM assuming test pass

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -4317,6 +4317,14 @@ Tensor linalg_det_backward(
     return {};
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp:4317]: Gradient is special cased here? Are you saying the special case is incorrect?

Comment (Skylion007) [reply torch/csrc/autograd/FunctionsManual.cpp:4317]: Oh yeah, the second or of the if statement is incorrect. Should be return zeros_like(A) then...

File: test/test_linalg.py [hunk 3]
```diff
@@ -1042,7 +1042,15 @@ def test_det(self, device, dtype):
         with self.assertRaises(RuntimeError):
             op(t)
 
-    @skipCUDAIfNoCusolver
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.double, torch.cdouble)
+    def test_det_backward(self):
```
Comment (Skylion007) [inline test/test_linalg.py]:
```suggestion
    def test_det_backward(self, device, dtype):
```

File: test/test_linalg.py [hunk 4]
```diff
@@ -1042,7 +1042,15 @@ def test_det(self, device, dtype):
...
+        # Regression test for #80761.
+        input = torch.tensor([[0.]], dtype=self.dtype, requires_grad=True)
```
Comment (Skylion007) [inline test/test_linalg.py]:
```suggestion
        input = torch.tensor([[0.]], device=device, dtype=self.dtype, requires_grad=True)

File: test/test_linalg.py [hunk 5]
```diff
@@ -1042,7 +1042,15 @@ def test_det(self, device, dtype):
         with self.assertRaises(RuntimeError):
             op(t)
 
-    @skipCUDAIfNoCusolver
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.double, torch.cdouble)
+    def test_det_backward(self, device, dtype):
+        # Regression test for #80761.
+        input = torch.tensor([[0.]], device=device, dtype=self.dtype, requires_grad=True)
```
Comment (Skylion007) [inline test/test_linalg.py]:
```suggestion
        input = torch.tensor([[0.]], device=device, dtype=dtype, requires_grad=True)
```
